p4include/psa.p4

```diff
@@ -0,0 +1,675 @@
+  Ingres2Egress,    /// post parse original ingres,  Buffering queue
+  Egress2Ingress,   /// post deparse in egress,      Ingress parser
+  Egress2Egress     /// inout to deparser in egress, Buffering queue
+}
+
+/* --------------------------------------------------------------------- */
+/*
+ * PSA Externs
+ */
+
+/**
+ * The PacketReplicationEngine extern represents the non-programmable
+ * part of the PSA pipeline.
+ *
+ * Even though the PRE can not be programmed using P4, it can be
+ * configured both directly using control plane APIs and by setting
+ * intrinsic metadata. In this specification we opt to define the
+ * operations available in the PRE as method invocations. A target
+ * backend is responsible for mapping the PRE extern APIs to the
+ * appropriate mechanisms for performing these operations in the
+ * hardware.
+ *
+ * The PRE is instantiated by the architecture and a P4 program can
+ * use it directly. It is an error to instantiate the PRE multiple
+ * times.  The PRE is made available to the Ingress programmable block
+ * using the same mechanism as packet_in. A corresponding Buffering
+ * and Queueing Engine (BQE) extern is defined for the Egress
+ * pipeline (@see BufferingQueueingEngine).
+ *
+ * Note: some of these operations may not be implemented as primitive
+ * operations on a certain target. However, All of the operations can
+ * be implemented as a combination of other operations. Applications
+ * that rely on non-primitive operations may incur significant
+ * performance penalties, however, they should be functionally
+ * correct.
+ *
+ * Semantics of behavior for multiple calls to PRE APIs
+ *
+ * The semantics of calling the PRE APIs is equivalent to setting
+ * intrinsic metadata fields/bits and assumming that the PRE looks up
+ * the fields in the following order: drop, truncate, multicast,
+ * clone, output_port.
+ *
+ * Following this semantics, examples of the behaviors are:
+ *
+ * - any call to drop in the pipeline will cause the packet, and all
+ *   potential clone copies (see below) to drop.
+ *
+ * - any call to truncate, will cause the packet (and its clones
+ *   \todo: check) to be truncated.
+ *
+ * - multiple calls to send_to_port -- the last call in the ingress
+ *   pipeline sets the output port.
+ *
+ * - multiple calls to multicast -- the last in the ingress pipeline
+ *   sets the multicast group
+ *
+ * - interleaving send_to_port and multicast -- the semantics of multicast
+ *   is defined as below (https://github.com/p4lang/tutorials/issues/22):
+ *     if (multicast_group != 0)
+ *        multicast_to_group(multicast_group);
+ *     else
+ *        send_to_port(output_port);
+ *
+ *   From this, it follows that if there is a call that sets the
+ *   multicast_group, the packet will be multicast to the group that
+ *   last set the multicast group. Otherwise, the packet will be sent
+ *   to the port set by either send_to_port.
```

Comment:
It is a textual representation of the if above, because the if does not actually capture the semantics of multiple calls. Not sure why you think it doesn't fit ...